src/v/kafka/client/partitioners.cc

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "kafka/client/partitioners.h"
+
+#include "hashing/murmur.h"
+
+#include <functional>
+#include <optional>
+#include <tuple>
+#include <type_traits>
+
+namespace kafka::client {
+
+namespace detail {
+
+class identity_partitioner final : public partitioner_impl {
+public:
+    std::optional<model::partition_id>
+    operator()(const record_essence& rec, size_t) override {
+        return rec.partition_id;
+    }
+};
+
+class murmur2_key_partitioner final : public partitioner_impl {
+public:
+    std::optional<model::partition_id>
+    operator()(const record_essence& rec, size_t partition_count) override {
+        if (!rec.key || rec.key->empty()) {
+            return std::nullopt;
+        }
+        iobuf_const_parser p(*rec.key);
+        auto key = p.read_bytes(p.bytes_left());
+        auto hash = murmur2(key.data(), key.size());
+        return model::partition_id(hash % partition_count);
+    }
+};
+
+class roundrobin_partitioner final : public partitioner_impl {
+public:
+    explicit roundrobin_partitioner(model::partition_id initial)
+      : partitioner_impl{}
+      , _next(initial) {}
+
+    std::optional<model::partition_id>
+    operator()(const record_essence&, size_t partition_count) override {
+        return model::partition_id(_next++ % partition_count);
+    }
+
+private:
+    model::partition_id _next;
+};
+
+template<typename... Impls>
+class composed_partitioner final : public partitioner_impl {
```

Comment:
default_partitioner?

Reply:
I've renamed `default_fallback_partitioner` as `default_partitioner`, and given `composed_partitioner` a comment. I think the name is fine, since it is a generic composition of a list of partitioner strategies.